plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts

```diff
@@ -94,16 +94,24 @@ export class HigherOrderOperations implements HigherOrderOperation {
     if (!previousLocation) {
       await this.locationsCatalog.addLocation(location);
     }
-    const outputEntities: Entity[] = [];
-    for (const entity of readerOutput.entities) {
-      const out = await this.entitiesCatalog.addOrUpdateEntity(
-        entity.entity,
-        location.id,
-      );
-      outputEntities.push(out);
+    if (readerOutput.entities.length === 0) {
+      return { location, entities: [] };
     }
 
-    return { location, entities: outputEntities };
+    const writtenEntities = await this.entitiesCatalog.batchAddOrUpdateEntities(
+      readerOutput.entities,
+      location.id,
+    );
+
+    if (writtenEntities.length === 0) {
```

Comment:
it's surprising that this can be anything but the exact same size as the input

Reply:
Yep can get rid of this